### ISSUE TITLE
Crash when dropping materialized view created in support of a secondary index

### ISSUE THREAD

**Issue by veramine** [scylladb/scylla#3357]
*Installation details*
Scylla version (or git commit hash):  2.1.0-0.20180216.58b90ceee
Cluster size: 1 
OS (RHEL/CentOS/Ubuntu/AWS AMI): AWS AMI

*Hardware details (for performance issues)*         
AWS i3.xlarge 

I had truncated a base table a few days ago that had 7 secondary indexes.  The base table at the time of truncatation had something like 30GB of data.  We had stopped using this table and were instead using a new table.  The base table truncation worked fine and I assumed that the materailized views would eventually be compacted and would also be truncated.  Today, I checked nodetool cfstats and saw that each of the MV's still held between 3-8GB's of data.  I tried fetching a few rows from one of them and the query timed out.  So I started dropping (not truncating) the MV's one at a time.  No data had been written to the base table for several days and it held 0 bytes at the time I attempted to drop its materialized views.  During the drop materialized view operation, the database crashed with the following backtrace:

```
/usr/src/debug/scylla-2.1.0/seastar/core/reactor.cc:294
/usr/src/debug/scylla-2.1.0/seastar/core/reactor.cc:322
/usr/src/debug/scylla-2.1.0/seastar/core/reactor.cc:3726
/lib64/libpthread.so.0+0x000000000000f5df
/usr/src/debug/scylla-2.1.0/utils/histogram.hh:49
/usr/src/debug/scylla-2.1.0/seastar/core/future.hh:1265
/usr/src/debug/scylla-2.1.0/seastar/core/future-util.hh:346
/usr/src/debug/scylla-2.1.0/seastar/core/reactor.cc:2431
/usr/src/debug/scylla-2.1.0/seastar/core/reactor.cc:2841
/opt/scylladb/include/c++/7/bits/unique_ptr.h:267
/usr/src/debug/scylla-2.1.0/seastar/core/reactor.cc:3931 (discriminator 3)
/opt/scylladb/include/c++/7/bits/std_function.h:706
/lib64/libpthread.so.0+0x0000000000007e24
/lib64/libc.so.6+0x00000000000f834c
```

Here was how we initially created the table and secondary indexes:

```
CREATE TABLE network_connections5 (
    customerguid uuid,
    year smallint,
    month tinyint,
    day tinyint,
    hour tinyint,
    starttstamp timeuuid,
    hostid int,
    processguid uuid,
    connectiontype int,
    flowid bigint,
    bindid bigint,
    processbasename text,
    pid int,
    unsignedprocess boolean,
    direction int,
    localaddress text,
    localport int,
    protocol int,
    remoteaddress text,
    remoteport int,
    remotednsname text,
    active boolean,
    endtime timestamp,
    duration bigint,
    datalengthin bigint,
    datalengthout bigint,
    PRIMARY KEY ((customerguid, year, month, day, hour), starttstamp))
        WITH CLUSTERING ORDER BY (starttstamp DESC)
        AND compaction = { 'class' : 'DateTieredCompactionStrategy' }
        AND default_time_to_live = 864000
        AND gc_grace_seconds = 30;
        
CREATE INDEX ON network_connections5 (remoteaddress);
CREATE INDEX ON network_connections5 (remoteport);
CREATE INDEX ON network_connections5 (remotednsname);
CREATE INDEX ON network_connections5 (processbasename);
CREATE INDEX ON network_connections5 (active);
CREATE INDEX ON network_connections5 (processguid);
CREATE INDEX ON network_connections5 (hostid);
```

Here is what showed up in journalctl at the time that I started dropping these materialized views one after another:

```
Apr 13 19:53:07 scylladb scylla[804]:  [shard 1] migration_manager - Drop view 'veraminetest.network_connections5_processguid_idx_index'
Apr 13 19:53:08 scylladb scylla[804]:  [shard 1] schema_tables - Dropping veraminetest.network_connections5_processguid_idx_index id=fdf6d4b
Apr 13 19:53:16 scylladb scylla[804]:  [shard 1] storage_proxy - Exception when communicating with 172.31.16.245: Can't find a column family
Apr 13 19:53:18 scylladb scylla[804]:  [shard 1] compaction - Compacting [/var/lib/scylla/data/system/local-7ad54392bcdd35a684174e047860b377
Apr 13 19:53:18 scylladb scylla[804]:  [shard 1] compaction - Compacted 4 sstables to [/var/lib/scylla/data/system/local-7ad54392bcdd35a6841
Apr 13 19:53:18 scylladb scylla[804]:  [shard 1] database - Schema version changed to 8a90ed65-b396-3caa-9c99-3b0552e9fc6f
Apr 13 19:53:36 scylladb scylla[804]:  [shard 1] compaction - Compacting [/var/lib/scylla/data/veraminetest/network_connections6_remoteaddre
Apr 13 19:53:39 scylladb scylla[804]:  [shard 1] migration_manager - Drop view 'veraminetest.network_connections5_remoteport_idx_index'
Apr 13 19:53:40 scylladb scylla[804]:  [shard 1] schema_tables - Dropping veraminetest.network_connections5_remoteport_idx_index id=fc725790
Apr 13 19:53:40 scylladb scylla[804]:  [shard 1] compaction - Compacting [/var/lib/scylla/data/system/local-7ad54392bcdd35a684174e047860b377
Apr 13 19:53:40 scylladb scylla[804]:  [shard 1] compaction - Compacted 4 sstables to [/var/lib/scylla/data/system/local-7ad54392bcdd35a6841
Apr 13 19:53:41 scylladb scylla[804]:  [shard 1] database - Schema version changed to 0450688a-5f20-3df8-8cc2-e9b1f32f6217
Apr 13 19:53:41 scylladb scylla[804]:  [shard 1] compaction - Compacted 4 sstables to [/var/lib/scylla/data/veraminetest/network_connections
Apr 13 19:53:53 scylladb scylla[804]:  [shard 1] migration_manager - Drop view 'veraminetest.network_connections5_remotednsname_idx_index'
Apr 13 19:53:53 scylladb scylla[804]:  [shard 2] compaction - Compacting [/var/lib/scylla/data/system_schema/columns-24101c25a2ae3af787c1b40
Apr 13 19:53:53 scylladb scylla[804]:  [shard 2] compaction - Compacting [/var/lib/scylla/data/system_schema/scylla_tables-5d912ff1f7593665b
Apr 13 19:53:53 scylladb scylla[804]:  [shard 2] compaction - Compacting [/var/lib/scylla/data/system_schema/views-9786ac1cdd583201a7cdad556
Apr 13 19:53:53 scylladb scylla[804]:  [shard 2] compaction - Compacted 4 sstables to [/var/lib/scylla/data/system_schema/scylla_tables-5d91
Apr 13 19:53:53 scylladb scylla[804]:  [shard 2] compaction - Compacting [/var/lib/scylla/data/system_schema/keyspaces-abac5682dea631c5b535b
Apr 13 19:53:53 scylladb scylla[804]:  [shard 2] compaction - Compacted 4 sstables to [/var/lib/scylla/data/system_schema/views-9786ac1cdd58
Apr 13 19:53:53 scylladb scylla[804]:  [shard 2] compaction - Compacted 4 sstables to [/var/lib/scylla/data/system_schema/columns-24101c25a2
Apr 13 19:53:54 scylladb scylla[804]:  [shard 2] compaction - Compacted 4 sstables to [/var/lib/scylla/data/system_schema/keyspaces-abac5682
Apr 13 19:53:54 scylladb scylla[804]:  [shard 1] schema_tables - Dropping veraminetest.network_connections5_remotednsname_idx_index id=fccfe
Apr 13 19:53:54 scylladb scylla[804]:  [shard 1] database - Schema version changed to d4fa9ef6-5d5f-3390-b699-7e060adfd9dc
Apr 13 19:54:04 scylladb scylla[804]:  [shard 1] migration_manager - Drop view 'veraminetest.network_connections5_processbasename_idx_index'
Apr 13 19:54:05 scylladb scylla[804]:  [shard 1] schema_tables - Dropping veraminetest.network_connections5_processbasename_idx_index id=fd3
Apr 13 19:54:05 scylladb scylla[804]:  [shard 1] compaction - Compacting [/var/lib/scylla/data/system/local-7ad54392bcdd35a684174e047860b377
Apr 13 19:54:05 scylladb scylla[804]:  [shard 1] compaction - Compacted 4 sstables to [/var/lib/scylla/data/system/local-7ad54392bcdd35a6841
Apr 13 19:54:05 scylladb scylla[804]:  [shard 1] database - Schema version changed to f56f7190-cc28-347f-ac72-d57f3600e488
Apr 13 19:54:18 scylladb scylla[804]:  [shard 1] migration_manager - Drop view 'veraminetest.network_connections5_active_idx_index'
Apr 13 19:54:18 scylladb scylla[804]:  [shard 1] schema_tables - Dropping veraminetest.network_connections5_active_idx_index id=fd9a83e0-1dd
Apr 13 19:54:18 scylladb scylla[804]: Segmentation fault on shard 3.
Apr 13 19:54:18 scylladb scylla[804]: Backtrace:
Apr 13 19:54:18 scylladb scylla[804]: 0x00000000004a38ab
Apr 13 19:54:18 scylladb scylla[804]: 0x00000000004a397c
Apr 13 19:54:18 scylladb scylla[804]: 0x00000000004a39d3
Apr 13 19:54:18 scylladb scylla[804]: /lib64/libpthread.so.0+0x000000000000f5df
Apr 13 19:54:18 scylladb scylla[804]: 0x0000000000b1b91e
Apr 13 19:54:18 scylladb scylla[804]: 0x0000000000b1cc19
Apr 13 19:54:18 scylladb scylla[804]: 0x0000000000b1ddc8
Apr 13 19:54:18 scylladb scylla[804]: 0x00000000004a026a
Apr 13 19:54:18 scylladb scylla[804]: 0x00000000004a054c
Apr 13 19:54:18 scylladb scylla[804]: 0x00000000004ef444
Apr 13 19:54:18 scylladb scylla[804]: 0x00000000004f4b32
Apr 13 19:54:18 scylladb scylla[804]: 0x00000000005577ed
Apr 13 19:54:18 scylladb scylla[804]: /lib64/libpthread.so.0+0x0000000000007e24
Apr 13 19:54:18 scylladb scylla[804]: /lib64/libc.so.6+0x00000000000f834c
Apr 13 19:55:27 scylladb scylla[688]: Scylla version 2.1.0-0.20180216.58b90ceee starting ...
Apr 13 19:55:27 scylladb scylla[688]:  [shard 0] init - Scylla API server listening on 127.0.0.1:10000 ...
...
```

Here are the scylla args we use to start the database.

```
SCYLLA_ARGS="--background-writer-scheduling-quota 0.5 --auto-adjust-flush-quota 1 --log-to-syslog 1 --log-to-stdout 0 --default-log-level info --collectd-address=127.0.0.1:25826 --collectd=1 --collectd-poll-period 3000 --network-stack posix --reserve-memory 6G"
```

I also have a core dump but its file size is 25,754,742,784 and would take a while to upload.  But let me know if you think it would be helpful to perform the investigation and I will do so.

Thanks.

**Comment by glommer** [comment]
@veramine decoding that backtrace - the same way you do for the stall traces - is a great start. Even without the core we can just with that at least pinpoint the culprit subsystem

**Comment by veramine** [comment]
I used the wrong addr2line options and wrong binary.  Sorry, here's the correct backtrace:

```
seastar::backtrace_buffer::append_backtrace() at /usr/src/debug/scylla-2.1.0/seastar/core/reactor.cc:294
 (inlined by) print_with_backtrace at /usr/src/debug/scylla-2.1.0/seastar/core/reactor.cc:315
seastar::print_with_backtrace(char const*) at /usr/src/debug/scylla-2.1.0/seastar/core/reactor.cc:322
void seastar::install_oneshot_signal_handler<11, &seastar::sigsegv_action>()::{lambda(int, siginfo_t*, void*)#1}::_FUN(int, siginfo_t*, void*) at /usr/src/debug/scylla-2.1.0/seastar/core/reactor.cc:3726
 (inlined by) operator() at /usr/src/debug/scylla-2.1.0/seastar/core/reactor.cc:3712
 (inlined by) _FUN at /usr/src/debug/scylla-2.1.0/seastar/core/reactor.cc:3708
?? ??:0
utils::moving_average::add(unsigned long) at /usr/src/debug/scylla-2.1.0/utils/histogram.hh:49
 (inlined by) utils::timed_rate_moving_average::mark(unsigned long) at /usr/src/debug/scylla-2.1.0/utils/histogram.hh:253
 (inlined by) row_cache::on_row_miss() at /usr/src/debug/scylla-2.1.0/row_cache.cc:459
 (inlined by) cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}::operator()(mutation_fragment) const at /usr/src/debug/scylla-2.1.0/cache_flat_mutation_reader.hh:270
 (inlined by) seastar::future<> consume_mutation_fragments_until<cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}&&)::{lambda()#2}::operator()() const at /usr/src/debug/scylla-2.1.0/flat_mutation_reader.hh:409
std::enable_if<seastar::is_future<std::result_of<seastar::future<> ()>::type>::value, {lambda()#3}>::type seastar::do_void_futurize_apply<seastar::future<> consume_mutation_fragments_until<cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}&&)::{lambda()#2}&>(flat_mutation_reader, seastar::is_future) at /usr/src/debug/scylla-2.1.0/seastar/core/future.hh:1265
 (inlined by) {lambda()#3} seastar::futurize<void>::apply<seastar::future<> consume_mutation_fragments_until<cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}&&)::{lambda()#2}&>(flat_mutation_reader) at /usr/src/debug/scylla-2.1.0/seastar/core/future.hh:1301
 (inlined by) {lambda()#3} seastar::do_until<seastar::future<> consume_mutation_fragments_until<cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}&&)::{lambda()#2}, {lambda()#3} consume_mutation_fragments_until<{lambda()#1}, mutation_fragment, {lambda(mutation_fragment)#2}>(seastar::future, flat_mutation_reader, {lambda()#1}, mutation_fragment)::{lambda()#1}>(flat_mutation_reader&, seastar::future<>) at /usr/src/debug/scylla-2.1.0/seastar/core/future-util.hh:342
{lambda()#3} seastar::do_until<seastar::future<> consume_mutation_fragments_until<cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}&&)::{lambda()#2}, {lambda()#3} consume_mutation_fragments_until<{lambda()#1}, mutation_fragment, {lambda(mutation_fragment)#2}>(seastar::future, flat_mutation_reader, {lambda()#1}, mutation_fragment)::{lambda()#1}>(flat_mutation_reader&, seastar::future<>)::{lambda()#1}::operator()() at /usr/src/debug/scylla-2.1.0/seastar/core/future-util.hh:346
 (inlined by) _ZN7seastar12apply_helperIZNS_8do_untilIZ32consume_mutation_fragments_untilIZN5cache26cache_flat_mutation_reader20read_from_underlyingEvEUlvE_ZNS4_20read_from_underlyingEvEUl17mutation_fragmentE0_ZNS4_20read_from_underlyingEvEUlvE1_ENS_6futureIJEEER20flat_mutation_readerOT_OT0_OT1_EUlvE0_ZS2_IS5_S7_S8_ESA_SC_SE_SG_SI_EUlvE_EESA_SF_SD_EUlvE_OSt5tupleIJEESt16integer_sequenceImJEEE5applyEOSL_SO_ at /usr/src/debug/scylla-2.1.0/seastar/core/apply.hh:36
 (inlined by) auto seastar::apply<{lambda()#3} seastar::do_until<seastar::future<> consume_mutation_fragments_until<cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}&&)::{lambda()#2}, {lambda()#3} consume_mutation_fragments_until<{lambda()#1}, mutation_fragment, {lambda(mutation_fragment)#2}>(seastar::future, flat_mutation_reader, {lambda()#1}, mutation_fragment)::{lambda()#1}>(flat_mutation_reader&, seastar::future<>)::{lambda()#1}>(flat_mutation_reader, std::tuple<>&&) at /usr/src/debug/scylla-2.1.0/seastar/core/apply.hh:44
 (inlined by) seastar::future<> seastar::futurize<seastar::future<> >::apply<seastar::future<> seastar::do_until<seastar::future<> consume_mutation_fragments_until<cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}&&)::{lambda()#2}, seastar::future<> consume_mutation_fragments_until<{lambda()#1}, mutation_fragment, {lambda(mutation_fragment)#2}>({lambda()#3}, flat_mutation_reader, {lambda()#1}, mutation_fragment)::{lambda()#1}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3})::{lambda()#1}>(flat_mutation_reader, std::tuple<>&&) at /usr/src/debug/scylla-2.1.0/seastar/core/future.hh:1308
 (inlined by) flat_mutation_reader& seastar::future<>::then<seastar::future<> seastar::do_until<seastar::future<> consume_mutation_fragments_until<cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}&&)::{lambda()#2}, seastar::future<> consume_mutation_fragments_until<{lambda()#1}, mutation_fragment, {lambda(mutation_fragment)#2}>({lambda()#3}, flat_mutation_reader, {lambda()#1}, mutation_fragment)::{lambda()#1}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3})::{lambda()#1}, seastar::future<> >(flat_mutation_reader)::{lambda(flat_mutation_reader)#1}::operator()<seastar::future_state<> >(auto, flat_mutation_reader) at /usr/src/debug/scylla-2.1.0/seastar/core/future.hh:905
 (inlined by) seastar::continuation<flat_mutation_reader& seastar::future<>::then<seastar::future<> seastar::do_until<seastar::future<> consume_mutation_fragments_until<cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}&&)::{lambda()#2}, seastar::future<> consume_mutation_fragments_until<{lambda()#1}, mutation_fragment, {lambda(mutation_fragment)#2}>({lambda()#3}, flat_mutation_reader, {lambda()#1}, mutation_fragment)::{lambda()#1}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3})::{lambda()#1}, seastar::future<> >(flat_mutation_reader)::{lambda(flat_mutation_reader)#1}>::run() at /usr/src/debug/scylla-2.1.0/seastar/core/future.hh:396
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at /usr/src/debug/scylla-2.1.0/seastar/core/reactor.cc:2431
seastar::reactor::run_some_tasks(std::chrono::time_point<std::chrono::_V2::steady_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >&) at /usr/src/debug/scylla-2.1.0/seastar/core/reactor.cc:2841
seastar::reactor::run_some_tasks(std::chrono::time_point<std::chrono::_V2::steady_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >&) at /opt/scylladb/include/c++/7/bits/unique_ptr.h:267
 (inlined by) seastar::reactor::run() at /usr/src/debug/scylla-2.1.0/seastar/core/reactor.cc:2985
seastar::smp::configure(boost::program_options::variables_map)::{lambda()#3}::operator()() const at /usr/src/debug/scylla-2.1.0/seastar/core/reactor.cc:3931 (discriminator 3)
std::function<void ()>::operator()() const at /opt/scylladb/include/c++/7/bits/std_function.h:706
 (inlined by) seastar::posix_thread::start_routine(void*) at /usr/src/debug/scylla-2.1.0/seastar/core/posix.cc:52
?? ??:0
?? ??:0
```

**Comment by veramine** [comment]
Here are the metrics and disk stats at the time of the crash.

![metrics1](https://user-images.githubusercontent.com/26036714/38758074-3cdd54a8-3f24-11e8-98e7-cf5bd1c5a108.png)

![disk1](https://user-images.githubusercontent.com/26036714/38758080-40befe46-3f24-11e8-97d2-38f8ff194ecd.png)

**Comment by veramine** [comment]
Glauber helped me extract the backtrace from the dump.  I will start it uploading tonight before I go to bed.

```
$ /opt/scylladb/bin/gdb /usr/bin/scylla /var/lib/scylla/coredump/core.scylla.994.bca1e6bf1bf14e18a4fb865a4a6c3453.804.1523649258000000 
GNU gdb (GDB) Red Hat Enterprise Linux 7.10.1-30.el7.centos
Copyright (C) 2015 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/bin/scylla...Reading symbols from /usr/lib/debug/usr/bin/scylla.debug...../../gdb/cp-support.c:1595: demangler-warning: unable to demangle '_ZN7seastar8futurizeI13frozen_schemaE5applyIRZNS_7shardedIN7service13storage_proxyEE9invoke_onIZZNS6_22init_messaging_serviceEvENKUljN5utils4UUIDEE6_clEjSA_EUlOT_E_NS_6futureIJS1_EEEEET0_jSD_EUlvE_JEEESG_SD_DpOT0_' (demangler failed with signal 11)
Unable to dump core, use `ulimit -c unlimited' before executing GDB next time.
../../gdb/cp-support.c:1608: demangler-warning: unable to demangle '_ZN7seastar8futurizeI13frozen_schemaE5applyIRZNS_7shardedIN7service13storage_proxyEE9invoke_onIZZNS6_22init_messaging_serviceEvENKUljN5utils4UUIDEE6_clEjSA_EUlOT_E_NS_6futureIJS1_EEEEET0_jSD_EUlvE_JEEESG_SD_DpOT0_' (demangler failed with signal 11)
A problem internal to GDB has been detected,
further debugging may prove unreliable.
Quit this debugging session? (y or n) n

This is a bug, please report it.  For instructions, see:
<http://www.gnu.org/software/gdb/bugs/>.

done.
done.
[New LWP 808]
[New LWP 814]
[New LWP 811]
[New LWP 809]
[New LWP 813]
[New LWP 816]
[New LWP 810]
[New LWP 815]
[New LWP 804]
[New LWP 807]
[New LWP 812]
[New LWP 806]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Core was generated by `/usr/bin/scylla --background-writer-scheduling-quota 0.5 --auto-adjust-flush-qu'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  cache_tracker::on_row_miss (this=0x3700275e7fc3a266) at row_cache.cc:188
188	    ++_stats.row_misses;
[Current thread is 1 (Thread 0x7f895bfff700 (LWP 808))]
Missing separate debuginfos, use: debuginfo-install bzip2-libs-1.0.6-13.el7.x86_64 cryptopp-5.6.2-10.el7.x86_64 elfutils-libelf-0.168-8.el7.x86_64 elfutils-libs-0.168-8.el7.x86_64 glibc-2.17-196.el7_4.2.x86_64 gmp-6.0.0-15.el7.x86_64 gnutls-3.3.26-9.el7.x86_64 gnutls-c++-3.3.26-9.el7.x86_64 hwloc-libs-1.11.2-2.el7.x86_64 jsoncpp-0.10.5-2.el7.x86_64 keyutils-libs-1.5.8-3.el7.x86_64 krb5-libs-1.15.1-8.el7.x86_64 libaio-0.3.109-13.el7.x86_64 libattr-2.4.46-12.el7.x86_64 libcap-2.22-9.el7.x86_64 libcom_err-1.42.9-10.el7.x86_64 libffi-3.0.13-18.el7.x86_64 libgcrypt-1.5.3-14.el7.x86_64 libgpg-error-1.12-3.el7.x86_64 libpciaccess-0.13.4-3.1.el7_4.x86_64 libselinux-2.5-11.el7.x86_64 libtasn1-4.10-1.el7.x86_64 libtool-ltdl-2.4.2-22.el7_3.x86_64 libunwind-1.2-2.el7.x86_64 libxml2-2.9.1-6.el7_2.3.x86_64 lz4-1.7.3-1.el7.x86_64 nettle-2.7.1-8.el7.x86_64 nss-softokn-freebl-3.28.3-8.el7_4.x86_64 numactl-libs-2.0.9-6.el7_2.x86_64 openssl-libs-1.0.2k-8.el7.x86_64 p11-kit-0.23.5-3.el7.x86_64 pcre-8.32-17.el7.x86_64 protobuf-2.5.0-8.el7.x86_64 scylla-libgcc72-7.2.1-1.el7.centos.x86_64 scylla-libstdc++72-7.2.1-1.el7.centos.x86_64 snappy-1.1.0-3.el7.x86_64 systemd-libs-219-42.el7_4.4.x86_64 thrift-0.9.1-15.el7.x86_64 xz-libs-5.2.2-1.el7.x86_64 yaml-cpp-0.5.1-6.el7.x86_64 zlib-1.2.7-17.el7.x86_64
(gdb) bt
#0  cache_tracker::on_row_miss (this=0x3700275e7fc3a266) at row_cache.cc:188
#1  row_cache::on_row_miss (this=<optimized out>) at row_cache.cc:460
#2  cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}::operator()(mutation_fragment) const (
    mf=<error reading variable: access outside bounds of object referenced via synthetic pointer>, __closure=0x7f895bfd4ca0) at cache_flat_mutation_reader.hh:270
#3  seastar::future<> consume_mutation_fragments_until<cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}&&)::{lambda()#2}::operator()() const (__closure=__closure@entry=0x7f895bfd4c90)
    at flat_mutation_reader.hh:409
#4  0x0000000000b1cc1a in seastar::do_void_futurize_apply<seastar::future<> consume_mutation_fragments_until<cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}&&)::{lambda()#2}&>(flat_mutation_reader, seastar::is_future) (func=...) at seastar/core/future.hh:1265
#5  seastar::futurize<void>::apply<seastar::future<> consume_mutation_fragments_until<cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}&&)::{lambda()#2}&>(flat_mutation_reader) (func=...)
    at seastar/core/future.hh:1301
#6  seastar::do_until<seastar::future<> consume_mutation_fragments_until<cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}&&)::{lambda()#2}, {lambda()#3} consume_mutation_fragments_until<{lambda()#1}, mutation_fragment, {lambda(mutation_fragment)#2}>(seastar::future, flat_mutation_reader, {lambda()#1}, mutation_fragment)::{lambda()#1}>(flat_mutation_reader&, seastar::future<>) (stop_cond=..., action=...) at seastar/core/future-util.hh:342
#7  0x0000000000b1ddc9 in {lambda()#3} seastar::do_until<seastar::future<> consume_mutation_fragments_until<cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}&&)::{lambda()#2}, {lambda()#3} consume_mutation_fragments_until<{lambda()#1}, mutation_fragment, {lambda(mutation_fragment)#2}>(seastar::future, flat_mutation_reader, {lambda()#1}, mutation_fragment)::{lambda()#1}>(flat_mutation_reader&, seastar::future<>)::{lambda()#1}::operator()() (__closure=<optimized out>) at seastar/core/future-util.hh:346
#8  seastar::apply_helper<{lambda()#3} seastar::do_until<seastar::future<> consume_mutation_fragments_until<cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}&&)::{lambda()#2}, {lambda()#3} consume_mutation_fragments_until<{lambda()#1}, mutation_fragment, {lambda(mutation_fragment)#2}>(seastar::future, flat_mutation_reader, {lambda()#1}, mutation_fragment)::{lambda()#1}>(flat_mutation_reader&, seastar::future<>)::{lambda()#1}, std::tuple<>&&, std::integer_sequence<unsigned long> >::apply (args=<optimized out>, 
    func=<optimized out>) at seastar/core/apply.hh:36
#9  seastar::apply<{lambda()#3} seastar::do_until<seastar::future<> consume_mutation_fragments_until<cache::cache_flat_mutation_reader::read_from_underlying()::{lam---Type <return> to continue, or q <return> to quit---
bda()#1}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}&&)::{lambda()#2}, {lambda()#3} consume_mutation_fragments_until<{lambda()#1}, mutation_fragment, {lambda(mutation_fragment)#2}>(seastar::future, flat_mutation_reader, {lambda()#1}, mutation_fragment)::{lambda()#1}>(flat_mutation_reader&, seastar::future<>)::{lambda()#1}>(flat_mutation_reader, std::tuple<>&&) (args=<optimized out>, func=<optimized out>) at seastar/core/apply.hh:44
#10 seastar::futurize<seastar::future<> >::apply<seastar::future<> seastar::do_until<seastar::future<> consume_mutation_fragments_until<cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}&&)::{lambda()#2}, seastar::future<> consume_mutation_fragments_until<{lambda()#1}, mutation_fragment, {lambda(mutation_fragment)#2}>({lambda()#3}, flat_mutation_reader, {lambda()#1}, mutation_fragment)::{lambda()#1}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3})::{lambda()#1}>(flat_mutation_reader, std::tuple<>&&) (args=<optimized out>, func=<unknown type in /usr/lib/debug/usr/bin/scylla.debug, CU 0x4c17806, DIE 0x4dd48bb>) at seastar/core/future.hh:1308
#11 flat_mutation_reader& seastar::future<>::then<seastar::future<> seastar::do_until<seastar::future<> consume_mutation_fragments_until<cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}&&)::{lambda()#2}, seastar::future<> consume_mutation_fragments_until<{lambda()#1}, mutation_fragment, {lambda(mutation_fragment)#2}>({lambda()#3}, flat_mutation_reader, {lambda()#1}, mutation_fragment)::{lambda()#1}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3})::{lambda()#1}, seastar::future<> >(flat_mutation_reader)::{lambda(flat_mutation_reader)#1}::operator()<seastar::future_state<> >(auto, flat_mutation_reader) (
    state=<unknown type in /usr/lib/debug/usr/bin/scylla.debug, CU 0x4c17806, DIE 0x4dd47ea>, __closure=0x603015d61918) at seastar/core/future.hh:905
#12 seastar::continuation<flat_mutation_reader& seastar::future<>::then<seastar::future<> seastar::do_until<seastar::future<> consume_mutation_fragments_until<cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#1}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda(mutation_fragment)#2}&&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3}&&)::{lambda()#2}, seastar::future<> consume_mutation_fragments_until<{lambda()#1}, mutation_fragment, {lambda(mutation_fragment)#2}>({lambda()#3}, flat_mutation_reader, {lambda()#1}, mutation_fragment)::{lambda()#1}>(flat_mutation_reader&, cache::cache_flat_mutation_reader::read_from_underlying()::{lambda()#3})::{lambda()#1}, seastar::future<> >(flat_mutation_reade
r)::{lambda(flat_mutation_reader)#1}>::run() (this=0x603015d61900) at seastar/core/future.hh:396
#13 0x00000000004a026b in seastar::reactor::run_tasks (this=this@entry=0x603000141000, tq=...) at core/reactor.cc:2431
#14 0x00000000004a054d in seastar::reactor::run_some_tasks (this=this@entry=0x603000141000, t_run_completed=...) at core/reactor.cc:2841
#15 0x00000000004ef445 in seastar::reactor::run_some_tasks (t_run_completed=..., this=0x603000141000) at /opt/scylladb/include/c++/7/bits/unique_ptr.h:267
#16 seastar::reactor::run (this=0x603000141000) at core/reactor.cc:2985
#17 0x00000000004f4b33 in seastar::smp::<lambda()>::operator()(void) const (__closure=0x6000000cd400) at core/reactor.cc:3931
#18 0x00000000005577ee in std::function<void ()>::operator()() const (this=<optimized out>) at /opt/scylladb/include/c++/7/bits/std_function.h:706
#19 seastar::posix_thread::start_routine (arg=<optimized out>) at core/posix.cc:52
#20 0x00007f8961128e25 in start_thread () from /lib64/libpthread.so.0
#21 0x00007f8960e5634d in clone () from /lib64/libc.so.6
(gdb) 
```

**Comment by tgrabiec** [comment]
I haven't looked at the coredump, but it smells like a use-after-free on the dropped `column_family`.

**Comment by nyh** [comment]
By the way, the problem of the MV tables not being truncated was already solved (see #3200), but of course we still need to fix the crash bug.

**Comment by duarten** [comment]
@tzach Why do you think this is related to materialized views and secondary indexes?

**Comment by tzach** [comment]
> @tzach Why do you think this is related to materialized views and secondary indexes?

From the issue subject

**Comment by nyh** [comment]
@duarten he got this problem when dropping a materialized view created automatically by the SI code (probably it's not a good thing that this is even allowed :-)). So it may be related to MV or SI - or just be a generic bug which was randomly found in an SI use case. I don't think we really know yet?

**Comment by duarten** [comment]
I think we probably have a pointer or reference where we would probably need to keep a `lw_shared_ptr`.

It's true that we should disallow `drop materialized view` on views backing secondary indexes, I've sent a patch for that.

**Comment by slivne** [comment]
@duarten did you fix this issue ?

**Comment by psarna** [comment]
I can't reproduce that, neither on master (with a slight modification to allow dropping mv backing secondary indexes) nor 58b90ceee. I used the original `network_connections5` schema and filled it with lots of data several times, but dropping indexes worked fine in both `debug` and `release`.

I see in the issue description that the table was truncated before, but 58b90ceee doesn't have #3200 fix 76e642391 applied, maybe that's what somehow caused the problem? I'm kind of out of ideas what to check.

**Comment by duarten** [comment]
Dropping indexes, or the MVs backing the indexes? 

If dropping the MVs in parallel with a stream of read/write operations (i.e., mixed cassandra-stress workload) doesn't trigger anything, then I guess we can close.

**Comment by psarna** [comment]
Yeah, sorry, MVs backing indexes, like in issue description. And I used various python scripts instead of cassandra-stress, but it was a stream of r/w operations anyway, in parallel with various combinations of dropping MVs that back indexes.

**Comment by duarten** [comment]
There's a core dump too, but I'm not sure that will tell use where the fiber started. 

Anyway, looking at the backtrace, it seems the error happened during a read, concurrent with dropping the view. There are two options:

1. This is a general problem, and we're not keeping a column family alive until all ongoing reads are done (which I'm not sure we do, or if so, how); or 
2. The issue was during the secondary index read path where we read from the view before reading from the base, where @nyh fixed some lifetime issues.

@tgrabiec Do you know if 1. is possible?

**Comment by tgrabiec** [comment]
@duarten We are not keeping `column_family` objects alive during reads. Reads started before drop, will proceed, reads started after drop, will fail. For regular reads at least it's not a problem that column_family goes away, mutation readers keep their `mutation_source`s (memtables, sstabes) and schema alive.

**Comment by duarten** [comment]
The cache's `snapshot_source` also captures the `column_family`. 

We should have reads use a phaser, or piggyback on the `_async_gate`, and wait on it during `database::drop_column_family`.

**Comment by tskweres** [comment]
Error while dropping materialized view on DB Table -> ResponseError: Cannot use DROP MATERIALIZED VIEW on Index

I'm using express-cassandra with ScyllaDB, I have my database set to 'drop' on each startup, I'm writing code / fleshing out my models.

When I start up the app, I get the error above. How can I just completely drop and rebuild the DB while I'm coding, and get rid of this error?

**Comment by psarna** [comment]
@kstweres materialized views which back secondary indexes should not be deleted directly - instead, you simply drop the base table and indexes are automatically dropped too (along with their underlying materialized views). The proper order for dropping a table with multiple materialized views and indexes would be:
1. Drop all materialized views manually created for this table with `CREATE MATERIALIZED VIEW` statement
2. Drop the table itself (its indexes will be dropped automatically as well)

**Comment by nyh** [comment]
@psarna I wonder if his problem is that existing code which lists materialized views lists the materialized views used for indexing. So existing code which wishes to remove all tables by removing all materialized views first fails. 
This is unfortunate (and arguably can be considered a bug in Cassandra compatibility and deserves a separate issue - I'll create one), but fixing such existing code should be easy: remove all the *indexes* from existing tables before attempting to list and remove the views.

**Comment by psarna** [comment]
@nyh yes, that's the root cause, and I think there already was a discussion or two whether we should include indexing-induced views to be listed as materialized views - I vaguely remember that we also already have such a github issue, but I can't find it now.

The same thing causes benign problems when people use the output of `DESCRIBE TABLE` to recreate a table (e.g. during migrations), since it also includes these automatically created views.

**Comment by nyh** [comment]
Opened https://github.com/scylladb/scylla/issues/6058

**Comment by nyh** [comment]
> @nyh yes, that's the root cause, and I think there already was a discussion or two whether we should include indexing-induced views to be listed as materialized views - I vaguely remember that we also already have such a github issue, but I can't find it now.

I also thought there would be such a github issue, but couldn't find one. So I created #6058, with two proposed solution. One difficult one is to fix the basic problem (that these views shouldn't be listed), but an easier way is just to make @tskweres 's application work, by allowing the view deletion.

**Comment by tskweres** [comment]
@nyh I appreciate you guys looking into this. As I have my database configured to 'drop' on every init, I'm assuming it should be able to drop on its own?

clientOptions: {
    contactPoints: process.env.SCYLLA_CONTACT.split(','),
    keyspace: process.env.SCYLLA_KEYSPACE,
    protocolOptions: {
      port: parseInt(process.env.SCYLLA_PORT, 10),
    },
    queryOptions: {
      consistency: 1,
    },
    authProvider: new auth.PlainTextAuthProvider(process.env.SCYLLA_USERNAME, process.env.SCYLLA_PASSWORD),
  },
  ormOptions: {
    createKeyspace: true,
    defaultReplicationStrategy: {
      class: 'SimpleStrategy',
      replication_factor: parseInt(process.env.SCYLLA_REPLICATION, 10),
    },
    migration: 'drop', // for development
  }
